# Damon Polioudakis
# 2015-12-03
# Compare metadata covariates to principal components of Luis 2015-12-02
# single-cell RNA-seq from VZ and CP

# Inputs
#   HTseq counts expression
#   Metadata
#   Read duplication percentage to add to metadata table

# Outputs
#   Correlation plot of principal compenents and covariates

rm(list=ls())
sessionInfo()

library(xlsx)
library(boot)
library(reshape2)

# Input data
# Gene expression from HTSeq
exDatDF <- read.csv("../SxaQSEQsXap089L2_HTSC/Exprs_HTSCexon.csv", row.names = 1)
# log2 FPMs generated by Jason Stein
lfpmDatDF <- read.csv("../data/FPM.csv")
lfpmDatDF <- lfpmDatDF[ ,-c(1, 3, 4, ncol(lfpmDatDF))]
# log2 FPMs normalized for cell cycle genes by Jason Stein
ccNormDatDF <- read.csv("../data/CellCycleNormalizedExpr.csv", row.names = 1)
# Metadata
metDatDF <- read.xlsx("../metadata/PercentageReadsMapping.xlsx", 1)
pctDupDF <- read.xlsx("../metadata/PercentDuplicates.xlsx", 1)
libBatch1DF <- read.xlsx("../metadata/LibraryBatch_Well_ID.xlsx", 1, header = FALSE)
libBatch2DF <- read.xlsx("../metadata/LibraryBatch_Well_ID.xlsx", 2, header = FALSE)
load("../analysis/tables/Gene_Length_GC_Bias.rda")
lenBiasDF <- data.frame(GeneLengthScore = lenBias)
gcBiasDF <- data.frame(GCcontent = gcBias)

# Output file path prefix
outPathCorr <- "../analysis/graphs/PCA_covariates_correlation"
################################################################################

# Format data sets

# Edit CellIDs to be same format
colnames(exDatDF) <- gsub("_.*$", "", colnames(exDatDF))
pctDupDF <- data.frame(PERCENT_DUPLICATION = t(pctDupDF)[-1, 1])
row.names(pctDupDF) <- gsub("\\..*\\.txt", "", row.names(pctDupDF))
# And add percent duplication data to metadata table
metDatDF <- merge(x = metDatDF, y = pctDupDF, by.x = "CellID", by.y = "row.names")
# Add batch ID to metadata
batc2ID <- gsub("^ID:2-", "", as.vector(t(libBatch2DF)))
batc2ID <- gsub("^ID:1-", "1_", batc2ID)
batc2ID <- gsub("_[[:digit:]].*$", "", batc2ID)
batc1ID <- gsub("^ID:2-", "", as.vector(t(libBatch1DF)))
batc1ID <- gsub("^ID:1-", "1_", batc1ID)
batc1ID <- gsub("_[[:digit:]].*$", "", batc1ID)
batchIDdF <- rbind(data.frame(LibraryBatch = "1", CaptureWellID = batc1ID)
                   , data.frame(LibraryBatch = "2", CaptureWellID = batc2ID))
metDatDF <- merge(x = metDatDF, y = batchIDdF,
                  by.x = "CaptureWellID", by.y = "CaptureWellID")
# Add length bias to metadata
row.names(lenBiasDF) <- gsub("_e.*", "", row.names(lenBiasDF))
lenBiasDF$CellID <- gsub("_e.*", "", row.names(lenBiasDF))
lenBiasDF <- lenBiasDF[2:(nrow(lenBiasDF)-1), ]
metDatDF <- merge(x = metDatDF, y = lenBiasDF, by.x = "CellID", by.y = "CellID")
# Add GC bias to metadata
row.names(gcBiasDF) <- gsub("_e.*", "", row.names(gcBiasDF))
gcBiasDF$CellID <- gsub("_e.*", "", row.names(gcBiasDF))
gcBiasDF <- gcBiasDF[2:(nrow(gcBiasDF)-1), ]
metDatDF <- merge(x = metDatDF, y = gcBiasDF, by.x = "CellID", by.y = "CellID")

# Filter metadata for cells included in FPM table from Jason Stein
lfpmCells <- gsub("X", "Cell", colnames(lfpmDatDF))
metLFPMdatDF <- metDatDF[metDatDF$CellID %in% lfpmCells, ]
lfpmDatDF <- lfpmDatDF[ ,order(colnames(lfpmDatDF))]

# Filter raw reads for genes > 5 count in at least 10% of samples
exNoERCCdF <- exDatDF[1:(nrow(exDatDF)-97), ]
keep <- apply(data.frame(exNoERCCdF > 5), 1, function(expd) sum(expd) >= 0.1*ncol(exNoERCCdF))
exF5Cin10SdF <- (exNoERCCdF[keep, ])

# Filter raw reads for genes > 5 count in at least 10% of samples and filter
# cells to the subset in Jason Stein's FPM data
# Use gene filter from above
lfpmCells <- gsub("X", "Cell", colnames(lfpmDatDF))
exF5Cin10SfiltCdF <- exF5Cin10SdF[ , colnames(exF5Cin10SdF) %in% lfpmCells]

# Filter metadata for cells included in cell cycle normalized FPM table from
# Jason Stein
ccNormCells <- gsub("X", "Cell", colnames(ccNormDatDF))
metCCnormDatDF <- metDatDF[metDatDF$CellID %in% ccNormCells, ]
ccNormDatDF <- ccNormDatDF[ ,order(colnames(ccNormDatDF))]

################################################################################

# PC Analysis

# Run PCA funtion
Run_PCA <- function (data, pClabel = "Unregressed") {
  pCdat <- prcomp(data, center=F);
  # pCdat <- prcomp(exDatDF, center=F);
  topPCs <- pCdat$rotation[,1:5];
  # Calculate variance explained by each PC
  varExp <- (pCdat$sdev)^2 / sum(pCdat$sdev^2)
  topVar <- varExp[1:5]
  colnames(topPCs) <- paste(pClabel, "\n", colnames(topPCs)
                            , " (", signif(100 * topVar[1:5], 2), "%)", sep = "")
  return(topPCs)
}

# Correlation panel function
panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor, ...) {
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  if (class(x) == "numeric" & class(y) == "numeric") {
    r <- abs(cor(x, y, use = "pairwise.complete.obs", method = "spearman"))
  } else {
    lmout <- lm(y~x)
    r <- sqrt(summary(lmout)$adj.r.squared)
  }
  txt <- format(c(r, 0.123456789), digits = digits)[1]
  txt <- paste0(prefix, txt)
  if(missing(cex.cor)) cex.cor <- 1.2/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r)
}

# Plot correlation panel function
Plot_Corr_Matrix <- function (pairsDat, topPCs, title, outPath) {
  pdf(outPath, height = 20, width = 24)
  pairs(cbind(pairsDat, topPCs), pch = 19
        , upper.panel = panel.cor
        , main = paste("Covariates and MaxQuant Comparison -- |Spearman's rho| correlation values",
                       "\n", title, sep = "")
  )
  dev.off()
}

# Select covariates
pairsDat <- metDatDF[c("uniquelymappedreads"
                       , "mappedtomultipleloci"
                       , "VisualQC"
                       , "CapturePlateID"
                       , "cDNA_Concentration"
                       , "CellCount"
                       , "NumReads"
                       , "PERCENT_DUPLICATION"
                       , "GeneLengthScore"
                       , "LibraryBatch"
                       , "GCcontent")]
pairsDat$PERCENT_DUPLICATION <- as.numeric(pairsDat$PERCENT_DUPLICATION)

# Select covariates for log2 FPM data
pairsLFPMdat <- metLFPMdatDF[c("uniquelymappedreads"
                       , "mappedtomultipleloci"
                       , "VisualQC"
                       , "CapturePlateID"
                       , "cDNA_Concentration"
                       , "CellCount"
                       , "NumReads"
                       , "PERCENT_DUPLICATION"
                       , "GeneLengthScore"
                       , "LibraryBatch"
                       , "GCcontent")]
pairsLFPMdat$PERCENT_DUPLICATION <- as.numeric(pairsLFPMdat$PERCENT_DUPLICATION)

# Select covariates for cell cycle normalized log2 FPM data
pairsCCnormDat <- metCCnormDatDF[c("uniquelymappedreads"
                               , "mappedtomultipleloci"
                               , "VisualQC"
                               , "CapturePlateID"
                               , "cDNA_Concentration"
                               , "CellCount"
                               , "NumReads"
                               , "PERCENT_DUPLICATION"
                               , "GeneLengthScore"
                               , "LibraryBatch"
                               , "GCcontent")]
pairsCCnormDat$PERCENT_DUPLICATION <- as.numeric(pairsCCnormDat$PERCENT_DUPLICATION)

# Run PCA

# Centers the mean of all genes - this means the PCA gives us the eigenvectors
# of the geneXgene covariance matrix, allowing us to assess the proportion of
# variance each component contributes to the data
meanCenteredM <- t(scale(log(t(as.matrix(exDatDF[1:(nrow(exDatDF)-97), ])+1), 2)
                         , scale=F))
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsDat, topPCs, title = "log2(reads + 1)"
                 , outPath = paste(outPathCorr, "_Log2readsP1.pdf", sep = ""))

meanCenteredM <- t(scale(t(as.matrix(exDatDF[1:(nrow(exDatDF)-97), ])), scale=F))
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsDat, topPCs, title = "Raw Reads"
                 , outPath = paste(outPathCorr, "_RawReads.pdf", sep = ""))

meanCenteredM <- as.matrix(exDatDF[1:(nrow(exDatDF)-97), ])
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsDat, topPCs, title = "No Mean Centering Raw Reads"
                 , outPath = paste(outPathCorr, "_NoMeanCentRawReads.pdf", sep = ""))

meanCenteredM <- log(as.matrix(exDatDF[1:(nrow(exDatDF)-97), ])+1, 2)
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsDat, topPCs, title = "No Mean Centering log2(reads + 1)"
                 , outPath = paste(outPathCorr, "_NoMeanCentLog2readsP1.pdf", sep = ""))

# PCA on log2 FPM subset of expressed genes
meanCenteredM <- t(scale(t(as.matrix(lfpmDatDF[, -1])), scale=F))
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsLFPMdat, topPCs
                 , title = "log2 FPM filtered for genes FPM > 5 in at least 10% of samples"
                 , outPath = paste(outPathCorr, "_Log2FPM_FG5in10S.pdf", sep = ""))

# PCA on log2 (reads +1) subset of expressed genes
meanCenteredM <- t(scale(log(t(as.matrix(exF5Cin10SdF)+2), 2), scale=F))
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsDat, topPCs
                 , title = "log2(reads + 1) filtered for genes count > 5 in at least 10% of samples"
                 , outPath = paste(outPathCorr, "_Log2readsP1_FG5in10S.pdf", sep = ""))

# PCA on log2 (reads +1) subset of expressed genes and subsets of cells in
# Jason Stein's FPM data table
meanCenteredM <- t(scale(log(t(as.matrix(exF5Cin10SfiltCdF)+1), 2), scale=F))
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsLFPMdat, topPCs
                 , title = "log2(reads + 1) filtered for genes count > 5 in at least 10% of samples
                 \nFiltered cells to subset in Jason's FPM data table"
                 , outPath = paste(outPathCorr, "_Log2readsP1_FG5in10S_FJScells.pdf", sep = ""))

# PCA on cell cycle normalized FPM from Jason Stein
meanCenteredM <- t(scale(t(as.matrix(ccNormDatDF)), scale=F))
topPCs <- Run_PCA(meanCenteredM)
Plot_Corr_Matrix(pairsCCnormDat, topPCs
                 , title = "Cell Cycle normalized FPM for genes FPM > 5 in at least 10% of samples"
                 , outPath = paste(outPathCorr, "_Log2FPM_FG5in10S_CCnorm.pdf", sep = ""))


# PCA on cell cycle normalized FPM from Jason Stein
# and PCA on log2 FPM subset of expressed genes
meanCenteredCCnormM <- t(scale(t(as.matrix(ccNormDatDF)), scale=F))
meanCenteredFPMm <- t(scale(t(as.matrix(lfpmDatDF[, -1])), scale=F))
topPCsCCnorm <- Run_PCA(meanCenteredCCnormM, pClabel = "CC Norm FPM")
topPCsFPM <- Run_PCA(meanCenteredFPMm, pClabel = "FPM")
row.names(topPCsFPM) <- gsub("X", "Cell", row.names(topPCsFPM))
topPCsFPM <- topPCsFPM[row.names(topPCsFPM) %in% row.names(topPCsCCnorm), ]
topPCs <- cbind(topPCsCCnorm, topPCsFPM)
Plot_Corr_Matrix(pairsCCnormDat, topPCs
                 , title = "Cell Cycle normalized FPM and FPM for genes FPM > 5 in at least 10% of samples"
                 , outPath = paste(outPathCorr, "_Log2FPM_FG5in10S_CCnorm_FPM.pdf", sep = ""))



